typing/ikinds/ikinds.ml

```diff
@@ -0,0 +1,1009 @@
+(* This forces ikinds globally on. *)
+Clflags.ikinds := true;
+(* CR jujacobs: set this to false before merging. *)
+
+(* Global feature toggles for the ikinds experiment.
+   These are intended to be easy to flip while iterating on
+   performance or correctness. *)
+(* CR jujacobs: remove toggles in the final version. *)
+Types.ikind_debug := false
+
+let enable_crossing = true
+
+let enable_sub_jkind_l = true
+
+let enable_sub_or_intersect = false
+
+let enable_sub_or_error = false
+
+let reset_constructor_ikind_on_substitution = false
+
+module Ldd = Ikind.Ldd
+
+module JK = struct
+  (* A JKind/ikind solver specialized to [Ikind.Ldd] and [Types.type_expr].
+
+     The solver computes LDD polynomials of the form
+       base ⊔ Σ_i (arg_i ⊓ coeff_i)
+     where [base] is the intrinsic kind of a constructor and each [coeff_i]
+     describes the contribution coming from the i-th type argument.
+     Least fixed points are used to interpret recursive types. *)
+
+  module Ldd = Ikind.Ldd
+
+  type ty = Types.type_expr
+
+  type constr = Ldd.constr
+
+  type lat = Ldd.lat
+
+  type poly = Ldd.node
+
+  type kind = Ldd.node
+
+  type mode =
+    | Normal
+    | Round_up
+
+  (* Hash tables avoiding polymorphic structural comparison on deep values.
+     We key types by their unique [Types.get_id] to keep lookups cheap. *)
+  module TyTbl = Hashtbl.Make (struct
+    type t = ty
+
+    let equal a b = Int.equal (Types.get_id a) (Types.get_id b)
+
+    let hash t = Hashtbl.hash (Types.get_id t)
+  end)
+
+  let constr_to_string (p : constr) : string = Format.asprintf "%a" Path.print p
+
+  module ConstrTbl = Hashtbl.Make (struct
+    type t = constr
+
+    let equal a b = Path.same a b
+
+    let hash x = Path.hash x
+  end)
```

Comment:
I suggest just writing `module Constr = Path` and defining a `Table` module in `path.ml`. Or even just writing `Path` everywhere rather than introducing a new alias. It's unclear to me whether this would be clearer, but it may be nice to avoid introducing new names for concepts used elsewhere in the compiler - up to you.